### PR TITLE
Adds a new event, "Radiation Leak"

### PR DESCRIPTION
**Pull request by MrMelbert**
## About The Pull Request

Adds a new random event, "Radiation Leak". 

A random machine in a random department will begin leaking radiation. 

This includes a radiation aura around it, and semi-regular puffs of smoke clouds containing mutagen and polonium.

The machine will stop leaking after some time, but can also be stopped by using a random tool on it, indicated in its examine text.

image

image

image

image

I added a new "radiation emitter" component, for simple "emits radiation" behavior. Then I made the radioactive mutation use it. This means that the radioactive gene emits radiation while on stasis and dead, rather than only in life. 

## Why It's Good For The Game

More randomly dangerous events to enforce the metal deathtrap theme.

## Changelog

:cl: Melbert
add: Added a new random event, "Radiation Leak". A random machine starts to emit radiation and toxic fumes. Engineers may be able to stop it with their tools. 
add: The "Radioactive" gene from genetics will now emit rads when on stasis or dead, rather than just while you're alive. 
/:cl:

### CODE REVIEW

File: code/datums/components/radioactive_emitter.dm [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	src.cooldown_time = max(cooldown_time, 0.25 SECONDS) // We'll cap it at every 4th of a second, we use fastprocess after all
+	src.range = range
+	src.threshold = threshold
+	src.examine_text = examine_text
+	src.signals_which_delete_us = signals_which_delete_us
+	src.sigreturn = sigreturn
+	src.on_signal_callback = on_signal_callback
+
+	// We process on fastprocess even though we're on a cooldown system, just so we're sure that we catch the next cooldown
+	START_PROCESSING(SSfastprocess, src)
```
Comment (LemonInTheDark) [inline code/datums/components/radioactive_emitter.dm:51]: Why not timers?

File: code/datums/components/radioactive_emitter.dm [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+/datum/component/radioactive_emitter/UnregisterFromParent()
+	if(length(signals_which_delete_us))
+		UnregisterSignal(parent, signals_which_delete_us)
+	UnregisterSignal(parent, COMSIG_PARENT_EXAMINE)
+
+/datum/component/radioactive_emitter/InheritComponent(
+	datum/component/new_comp,
+	i_am_original,
+	processing_subsystem = SSprocessing,
```
Comment (LemonInTheDark) [inline code/datums/components/radioactive_emitter.dm:73]: what is this arg for?

File: code/datums/components/radioactive_emitter.dm [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	SIGNAL_HANDLER
+
+	if(on_signal_callback)
+		INVOKE_ASYNC(src, PROC_REF(aync_delete_us), args)
+	else
+		qdel(src)
+
+	return sigreturn
+
+/datum/component/radioactive_emitter/proc/aync_delete_us(list/sig_args)
```
Comment (LemonInTheDark) [inline code/datums/components/radioactive_emitter.dm:110]: why is it called aync (async) delete us? what's async about it?

File: code/modules/events/radiation_leak.dm [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	start_when = 1 // 2 seconds in
+	announce_when = 10 // 20 seconds in
+	end_when = 150 // 300 seconds / ~5 minutes in
+
+	var/datum/weakref/picked_machine
+
+/datum/round_event/radiation_leak/setup()
+	// Pick a xeno spawn somewhere in the world.
+	// We will try to find a machine within a few turfs of it to start spewing rads from.
+	var/list/possible_locs = GLOB.generic_event_spawns.Copy()
```
Comment (LemonInTheDark) [inline code/modules/events/radiation_leak.dm:20]: why go off xeno spawns? that doesn't seem like the right thing

File: code/modules/events/radiation_leak.dm [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+			// Skip undertiles
+			if(sick_device.IsObscured())
+				continue
+			// Skip invisible stuff
+			if(sick_device.invisibility || !sick_device.alpha || !sick_device.mouse_opacity)
+				continue
+			// Look for dense machinery. Basically stops stuff like wall mounts and pipes, silly ones.
+			// But keep in vents and scrubbers. I think it's funny if they start spitting out radiation
+			if(!sick_device.density && !istype(sick_device, /obj/machinery/atmospherics/components/unary))
+				continue
```
Comment (LemonInTheDark) [inline code/modules/events/radiation_leak.dm:33]: you should do the cheapest/most common checks first
so density/pipes, invisible things, then obscured

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/modules/events/radiation_leak.dm [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+			// Skip invisible stuff
+			if(sick_device.invisibility || !sick_device.alpha || !sick_device.mouse_opacity)
+				continue
+			// Skip undertiles
+			if(sick_device.IsObscured())
+				continue
+			// Very basic check for atmos passability here
+			// We don't want to put our smoke inside something that can't spread it out, like airlocks
+			if(sick_device.can_atmos_pass() != ATMOS_PASS_YES)
+				continue
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/modules/events/radiation_leak.dm:40]: what exactly do you want to check for? at least one atmos transferable turf? `TURF_SHARES(turf)`?

[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/modules/events/radiation_leak.dm:40]: Alright. Can use turf_shares to ensure at least some spreading happens, up to you

**Comment by the-orange-cow** [comment]
Possibly outside the scope of this pull, but it would be cool if sabotaging a machine to do this was some sort of traitor side objective or something.

**Comment by Pumpkinoe** [comment]
I love that this implies that every machine in the station is at least partially radioactive
I also love that this might mean we get radioactive coffee machines

**Comment by Shadowflame909** [comment]
Engineers and Plasma-Men content. We love to see it.  My poor plasma-men coworker who I make fix it every time might not though

**Comment by MrDoomBringer** [comment]
hilarious and soulful, great change

**Comment by Jackal-boop** [comment]
> yes, it's even shown in one of the pictures in the pr

I am blind

**Comment by tralezab** [comment]
image
I get that radiation is green but greentext being success text is such a pattern that it makes me feel "YES! THE RADIATION IS LEAKING! RADBROS WE WON" when I read that

**Pull request merged**